common/render_shaders.go

```diff
@@ -783,6 +775,320 @@ func (l *legacyShader) Post() {
+	}
+	if changed := l.generateBufferContent(ren, space, ren.bufferContent); !changed {
+		return
+	}
+
+	if ren.buffer == nil {
+		ren.buffer = engo.Gl.CreateBuffer()
+	}
+	engo.Gl.BindBuffer(engo.Gl.ARRAY_BUFFER, ren.buffer)
+	engo.Gl.BufferData(engo.Gl.ARRAY_BUFFER, ren.bufferContent, engo.Gl.STATIC_DRAW)
+}
+
+func (l *textShader) generateBufferContent(ren *RenderComponent, space *SpaceComponent, buffer []float32) bool {
+	var changed bool
+
+	tint := colorToFloat32(ren.Color)
+	txt, ok := ren.Drawable.(Text)
+	if !ok {
+		unsupportedType(ren.Drawable)
+		return false
+	}
+
+	atlas, ok := atlasCache[txt.Font.URL]
+	if !ok {
+		// Generate texture first
+		atlas = txt.Font.generateFontAtlas(UnicodeCap)
+		atlasCache[txt.Font.URL] = atlas
+	}
+
+	var currentX float32
+	var currentY float32
+
+	var modifier float32 = 1
+	if txt.RightToLeft {
+		modifier = -1
+	}
+
+	for index, char := range txt.Text {
+		// TODO: this might not work for all characters
+		switch {
+		case char == '\n':
+			currentX = 0
+			currentY += atlas.Height[char] + txt.LineSpacing*atlas.Height[char]
+			continue
+		case char < 32: // all system stuff should be ignored
+			continue
+		}
+
+		offset := 20 * index
+
+		// These five are at 0, 0:
+		setBufferValue(buffer, 0+offset, currentX, &changed)
+		setBufferValue(buffer, 1+offset, currentY, &changed)
+		setBufferValue(buffer, 2+offset, atlas.XLocation[char]/atlas.TotalWidth, &changed)
+		setBufferValue(buffer, 3+offset, atlas.YLocation[char]/atlas.TotalHeight, &changed)
+		setBufferValue(buffer, 4+offset, tint, &changed)
+
+		// These five are at 1, 0:
+		setBufferValue(buffer, 5+offset, currentX+atlas.Width[char], &changed)
+		setBufferValue(buffer, 6+offset, currentY, &changed)
+		setBufferValue(buffer, 7+offset, (atlas.XLocation[char]+atlas.Width[char])/atlas.TotalWidth, &changed)
+		setBufferValue(buffer, 8+offset, atlas.YLocation[char]/atlas.TotalHeight, &changed)
+		setBufferValue(buffer, 9+offset, tint, &changed)
+
+		// These five are at 1, 1:
+		setBufferValue(buffer, 10+offset, currentX+atlas.Width[char], &changed)
+		setBufferValue(buffer, 11+offset, currentY+atlas.Height[char], &changed)
+		setBufferValue(buffer, 12+offset, (atlas.XLocation[char]+atlas.Width[char])/atlas.TotalWidth, &changed)
+		setBufferValue(buffer, 13+offset, (atlas.YLocation[char]+atlas.Height[char])/atlas.TotalHeight, &changed)
+		setBufferValue(buffer, 14+offset, tint, &changed)
+
+		// These five are at 0, 1:
+		setBufferValue(buffer, 15+offset, currentX, &changed)
+		setBufferValue(buffer, 16+offset, currentY+atlas.Height[char], &changed)
+		setBufferValue(buffer, 17+offset, atlas.XLocation[char]/atlas.TotalWidth, &changed)
+		setBufferValue(buffer, 18+offset, (atlas.YLocation[char]+atlas.Height[char])/atlas.TotalHeight, &changed)
+		setBufferValue(buffer, 19+offset, tint, &changed)
+
+		currentX += modifier * (atlas.Width[char] + float32(txt.Font.Size)*txt.LetterSpacing)
+	}
+
+	return changed
+}
```

Comment:
This is the "most important" part of the shader: it tells OpenGL for each corner, where it should go. The first two (out of 5), give the location relative to the origin of the object. The two after that, give the location relative to the FontAtlas (value between 0 and 1 to indicate where on that FontAtlas image the character is drawn). The last one gives the `tint` / color-filter.